{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Jaculus_jaculus protein</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Ref</th>\n",
       "      <th>Alt</th>\n",
       "      <th>Identity</th>\n",
       "      <th>Orthologous protein</th>\n",
       "      <th>PROVEAN</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSJJAP00000010175</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>R</td>\n",
       "      <td>H</td>\n",
       "      <td>0.494</td>\n",
       "      <td>ENSRNOP00000074565</td>\n",
       "      <td>4.606016</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSJJAP00000010175</td>\n",
       "      <td>7</td>\n",
       "      <td>7</td>\n",
       "      <td>D</td>\n",
       "      <td>N</td>\n",
       "      <td>0.494</td>\n",
       "      <td>ENSRNOP00000074565</td>\n",
       "      <td>1.492671</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSJJAP00000010175</td>\n",
       "      <td>10</td>\n",
       "      <td>10</td>\n",
       "      <td>E</td>\n",
       "      <td>D</td>\n",
       "      <td>0.494</td>\n",
       "      <td>ENSRNOP00000074565</td>\n",
       "      <td>4.676087</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSJJAP00000010175</td>\n",
       "      <td>12</td>\n",
       "      <td>12</td>\n",
       "      <td>S</td>\n",
       "      <td>R</td>\n",
       "      <td>0.494</td>\n",
       "      <td>ENSRNOP00000074565</td>\n",
       "      <td>-7.191513</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSJJAP00000010175</td>\n",
       "      <td>21</td>\n",
       "      <td>21</td>\n",
       "      <td>R</td>\n",
       "      <td>W</td>\n",
       "      <td>0.494</td>\n",
       "      <td>ENSRNOP00000074565</td>\n",
       "      <td>4.732174</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  Jaculus_jaculus protein  Start  End Ref Alt  Identity Orthologous protein  \\\n",
       "0      ENSJJAP00000010175      3    3   R   H     0.494  ENSRNOP00000074565   \n",
       "1      ENSJJAP00000010175      7    7   D   N     0.494  ENSRNOP00000074565   \n",
       "2      ENSJJAP00000010175     10   10   E   D     0.494  ENSRNOP00000074565   \n",
       "3      ENSJJAP00000010175     12   12   S   R     0.494  ENSRNOP00000074565   \n",
       "4      ENSJJAP00000010175     21   21   R   W     0.494  ENSRNOP00000074565   \n",
       "\n",
       "    PROVEAN  \n",
       "0  4.606016  \n",
       "1  1.492671  \n",
       "2  4.676087  \n",
       "3 -7.191513  \n",
       "4  4.732174  "
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = pd.read_table('/Users/hq/Desktop/total_var.tsv', header=0)\n",
    "df.head(5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(38973, 8)"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "small_df = df[(df['Identity'] > 0.6) & (df['PROVEAN'] < -12)]\n",
    "small_df.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "5000\n",
      "10000\n",
      "15000\n",
      "20000\n",
      "25000\n",
      "30000\n",
      "35000\n"
     ]
    }
   ],
   "source": [
    "# df = df.head(5000)\n",
    "n = 0\n",
    "def f(x):\n",
    "    global n\n",
    "    n += 1\n",
    "    if n % 5000 == 0:\n",
    "        print(n)\n",
    "    return pd.Series(dict(Identity = ':'.join(map(lambda i: '%.3f' % i, x['Identity'].tolist())),\n",
    "                          Orthologous_protein = ':'.join(x['Orthologous protein'].tolist()), \n",
    "                          PROVEAN = x['PROVEAN'].mean()))\n",
    "\n",
    "merged_df = small_df.groupby(['Jaculus_jaculus protein', 'Start', 'End', 'Ref', 'Alt']).apply(f)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "merged_df.to_csv('/Users/hq/Desktop/total_var.filtered.tsv', sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>Identity</th>\n",
       "      <th>Orthologous_protein</th>\n",
       "      <th>PROVEAN</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Jaculus_jaculus protein</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Ref</th>\n",
       "      <th>Alt</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th rowspan=\"3\" valign=\"top\">ENSJJAP00000000001</th>\n",
       "      <th>45</th>\n",
       "      <th>45</th>\n",
       "      <th>L</th>\n",
       "      <th>I</th>\n",
       "      <td>0.783</td>\n",
       "      <td>ENSMAUP00000000001</td>\n",
       "      <td>-13.493077</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>102</th>\n",
       "      <th>102</th>\n",
       "      <th>I</th>\n",
       "      <th>M</th>\n",
       "      <td>0.789</td>\n",
       "      <td>ENSRNOP00000049172</td>\n",
       "      <td>-12.134189</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>312</th>\n",
       "      <th>313</th>\n",
       "      <th>SF</th>\n",
       "      <th>AS</th>\n",
       "      <td>0.783</td>\n",
       "      <td>ENSMAUP00000000001</td>\n",
       "      <td>-12.151438</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSJJAP00000000004</th>\n",
       "      <th>21</th>\n",
       "      <th>21</th>\n",
       "      <th>T</th>\n",
       "      <th>M</th>\n",
       "      <td>0.912</td>\n",
       "      <td>ENSMAUP00000000004</td>\n",
       "      <td>-13.131403</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"3\" valign=\"top\">ENSJJAP00000000006</th>\n",
       "      <th>7</th>\n",
       "      <th>7</th>\n",
       "      <th>A</th>\n",
       "      <th>S</th>\n",
       "      <td>0.832</td>\n",
       "      <td>ENSCGRP00000000006</td>\n",
       "      <td>-12.875577</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <th>21</th>\n",
       "      <th>VII</th>\n",
       "      <th>IVV</th>\n",
       "      <td>0.810</td>\n",
       "      <td>ENSMOCP00000000006</td>\n",
       "      <td>-12.461650</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>45</th>\n",
       "      <th>45</th>\n",
       "      <th>L</th>\n",
       "      <th>F</th>\n",
       "      <td>0.805</td>\n",
       "      <td>ENSMAUP00000000006</td>\n",
       "      <td>-17.088148</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"3\" valign=\"top\">ENSJJAP00000000010</th>\n",
       "      <th>113</th>\n",
       "      <th>113</th>\n",
       "      <th>A</th>\n",
       "      <th>T</th>\n",
       "      <td>0.706</td>\n",
       "      <td>ENSRNOP00000043141</td>\n",
       "      <td>-13.320975</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>230</th>\n",
       "      <th>230</th>\n",
       "      <th>V</th>\n",
       "      <th>I</th>\n",
       "      <td>0.715</td>\n",
       "      <td>ENSCGRP00000000010</td>\n",
       "      <td>-12.235830</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>317</th>\n",
       "      <th>317</th>\n",
       "      <th>M</th>\n",
       "      <th>I</th>\n",
       "      <td>0.717</td>\n",
       "      <td>ENSMOCP00000000010</td>\n",
       "      <td>-12.573487</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                          Identity Orthologous_protein  \\\n",
       "Jaculus_jaculus protein Start End Ref Alt                                \n",
       "ENSJJAP00000000001      45    45  L   I      0.783  ENSMAUP00000000001   \n",
       "                        102   102 I   M      0.789  ENSRNOP00000049172   \n",
       "                        312   313 SF  AS     0.783  ENSMAUP00000000001   \n",
       "ENSJJAP00000000004      21    21  T   M      0.912  ENSMAUP00000000004   \n",
       "ENSJJAP00000000006      7     7   A   S      0.832  ENSCGRP00000000006   \n",
       "                        19    21  VII IVV    0.810  ENSMOCP00000000006   \n",
       "                        45    45  L   F      0.805  ENSMAUP00000000006   \n",
       "ENSJJAP00000000010      113   113 A   T      0.706  ENSRNOP00000043141   \n",
       "                        230   230 V   I      0.715  ENSCGRP00000000010   \n",
       "                        317   317 M   I      0.717  ENSMOCP00000000010   \n",
       "\n",
       "                                             PROVEAN  \n",
       "Jaculus_jaculus protein Start End Ref Alt             \n",
       "ENSJJAP00000000001      45    45  L   I   -13.493077  \n",
       "                        102   102 I   M   -12.134189  \n",
       "                        312   313 SF  AS  -12.151438  \n",
       "ENSJJAP00000000004      21    21  T   M   -13.131403  \n",
       "ENSJJAP00000000006      7     7   A   S   -12.875577  \n",
       "                        19    21  VII IVV -12.461650  \n",
       "                        45    45  L   F   -17.088148  \n",
       "ENSJJAP00000000010      113   113 A   T   -13.320975  \n",
       "                        230   230 V   I   -12.235830  \n",
       "                        317   317 M   I   -12.573487  "
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "merged_df.head(10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Jaculus_jaculus protein</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Ref</th>\n",
       "      <th>Alt</th>\n",
       "      <th>Identity</th>\n",
       "      <th>Orthologous_protein</th>\n",
       "      <th>PROVEAN</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSJJAP00000000001</td>\n",
       "      <td>45</td>\n",
       "      <td>45</td>\n",
       "      <td>L</td>\n",
       "      <td>I</td>\n",
       "      <td>0.783</td>\n",
       "      <td>ENSMAUP00000000001</td>\n",
       "      <td>-13.493077</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSJJAP00000000001</td>\n",
       "      <td>102</td>\n",
       "      <td>102</td>\n",
       "      <td>I</td>\n",
       "      <td>M</td>\n",
       "      <td>0.789</td>\n",
       "      <td>ENSRNOP00000049172</td>\n",
       "      <td>-12.134189</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSJJAP00000000001</td>\n",
       "      <td>312</td>\n",
       "      <td>313</td>\n",
       "      <td>SF</td>\n",
       "      <td>AS</td>\n",
       "      <td>0.783</td>\n",
       "      <td>ENSMAUP00000000001</td>\n",
       "      <td>-12.151438</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSJJAP00000000004</td>\n",
       "      <td>21</td>\n",
       "      <td>21</td>\n",
       "      <td>T</td>\n",
       "      <td>M</td>\n",
       "      <td>0.912</td>\n",
       "      <td>ENSMAUP00000000004</td>\n",
       "      <td>-13.131403</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSJJAP00000000006</td>\n",
       "      <td>7</td>\n",
       "      <td>7</td>\n",
       "      <td>A</td>\n",
       "      <td>S</td>\n",
       "      <td>0.832</td>\n",
       "      <td>ENSCGRP00000000006</td>\n",
       "      <td>-12.875577</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  Jaculus_jaculus protein  Start  End Ref Alt Identity Orthologous_protein  \\\n",
       "0      ENSJJAP00000000001     45   45   L   I    0.783  ENSMAUP00000000001   \n",
       "1      ENSJJAP00000000001    102  102   I   M    0.789  ENSRNOP00000049172   \n",
       "2      ENSJJAP00000000001    312  313  SF  AS    0.783  ENSMAUP00000000001   \n",
       "3      ENSJJAP00000000004     21   21   T   M    0.912  ENSMAUP00000000004   \n",
       "4      ENSJJAP00000000006      7    7   A   S    0.832  ENSCGRP00000000006   \n",
       "\n",
       "     PROVEAN  \n",
       "0 -13.493077  \n",
       "1 -12.134189  \n",
       "2 -12.151438  \n",
       "3 -13.131403  \n",
       "4 -12.875577  "
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = pd.read_table('/Users/hq/Desktop/total_var.filtered.tsv')\n",
    "df.head(5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [],
   "source": [
    "counter_dict = {}\n",
    "for p in df['Orthologous_protein'].tolist():\n",
    "    pl = p.split(':')\n",
    "    for pp in pl:\n",
    "        try:\n",
    "            counter_dict[pp] += 1\n",
    "        except KeyError:\n",
    "            counter_dict[pp] = 1\n",
    "    \n",
    "counter_series = pd.Series(counter_dict)\n",
    "\n",
    "counts = df['Jaculus_jaculus protein'].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "metadata": {},
   "outputs": [],
   "source": [
    "results = pd.cut(counts, bins=[0, 1, 2, 3, 5, 10, 25, 1000], right=True, labels=['1', '2', '3', '4-5', '6-10', '11-25', '>25']).value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[3920, 2677, 1791, 1790, 1299, 376, 23]"
      ]
     },
     "execution_count": 62,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "results.tolist()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_id = pd.read_table('jj_gene_protein.txt', sep='\\t', header=0, index_col='Protein stable ID')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Protein stable ID\n",
       "ENSJJAP00000000001    ENSJJAG00000000006\n",
       "ENSJJAP00000000002    ENSJJAG00000000010\n",
       "ENSJJAP00000000003    ENSJJAG00000000016\n",
       "ENSJJAP00000000004    ENSJJAG00000000019\n",
       "ENSJJAP00000000005    ENSJJAG00000000021\n",
       "ENSJJAP00000000006    ENSJJAG00000000022\n",
       "ENSJJAP00000000007    ENSJJAG00000000023\n",
       "ENSJJAP00000000008    ENSJJAG00000000025\n",
       "ENSJJAP00000000009    ENSJJAG00000000027\n",
       "ENSJJAP00000000010    ENSJJAG00000000028\n",
       "ENSJJAP00000000011    ENSJJAG00000000032\n",
       "ENSJJAP00000000012    ENSJJAG00000000033\n",
       "ENSJJAP00000000013    ENSJJAG00000000035\n",
       "ENSJJAP00000000016    ENSJJAG00000000038\n",
       "ENSJJAP00000000015    ENSJJAG00000000039\n",
       "ENSJJAP00000000014    ENSJJAG00000000040\n",
       "ENSJJAP00000000017    ENSJJAG00000000041\n",
       "ENSJJAP00000000018    ENSJJAG00000000042\n",
       "ENSJJAP00000000019    ENSJJAG00000000043\n",
       "ENSJJAP00000000020    ENSJJAG00000000043\n",
       "ENSJJAP00000000023    ENSJJAG00000000044\n",
       "ENSJJAP00000000022    ENSJJAG00000000046\n",
       "ENSJJAP00000000021    ENSJJAG00000000047\n",
       "ENSJJAP00000000025    ENSJJAG00000000048\n",
       "ENSJJAP00000000024    ENSJJAG00000000049\n",
       "ENSJJAP00000000026    ENSJJAG00000000050\n",
       "ENSJJAP00000000027    ENSJJAG00000000051\n",
       "ENSJJAP00000000028    ENSJJAG00000000052\n",
       "ENSJJAP00000000029    ENSJJAG00000000052\n",
       "ENSJJAP00000000030    ENSJJAG00000000053\n",
       "                             ...        \n",
       "ENSJJAP00000025151    ENSJJAG00000024411\n",
       "ENSJJAP00000025147    ENSJJAG00000024412\n",
       "ENSJJAP00000025154    ENSJJAG00000024413\n",
       "ENSJJAP00000025155    ENSJJAG00000024413\n",
       "ENSJJAP00000025156    ENSJJAG00000024414\n",
       "ENSJJAP00000025157    ENSJJAG00000024415\n",
       "ENSJJAP00000025159    ENSJJAG00000024416\n",
       "ENSJJAP00000025161    ENSJJAG00000024416\n",
       "ENSJJAP00000025162    ENSJJAG00000024416\n",
       "ENSJJAP00000025164    ENSJJAG00000024416\n",
       "ENSJJAP00000025158    ENSJJAG00000024417\n",
       "ENSJJAP00000025160    ENSJJAG00000024418\n",
       "ENSJJAP00000025163    ENSJJAG00000024418\n",
       "ENSJJAP00000025165    ENSJJAG00000024419\n",
       "ENSJJAP00000025166    ENSJJAG00000024420\n",
       "ENSJJAP00000025167    ENSJJAG00000024422\n",
       "ENSJJAP00000025168    ENSJJAG00000024423\n",
       "ENSJJAP00000025169    ENSJJAG00000024424\n",
       "ENSJJAP00000025170    ENSJJAG00000024425\n",
       "ENSJJAP00000025171    ENSJJAG00000024426\n",
       "ENSJJAP00000025172    ENSJJAG00000024427\n",
       "ENSJJAP00000025174    ENSJJAG00000024428\n",
       "ENSJJAP00000025175    ENSJJAG00000024428\n",
       "ENSJJAP00000025173    ENSJJAG00000024429\n",
       "ENSJJAP00000025176    ENSJJAG00000024430\n",
       "ENSJJAP00000025177    ENSJJAG00000024431\n",
       "ENSJJAP00000025178    ENSJJAG00000024432\n",
       "ENSJJAP00000025179    ENSJJAG00000024433\n",
       "ENSJJAP00000025180    ENSJJAG00000024434\n",
       "ENSJJAP00000025181    ENSJJAG00000024434\n",
       "Name: Gene stable ID, Length: 25180, dtype: object"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_id['Gene stable ID']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "SingleLetterAlphabet() alignment with 6 rows and 40 columns\n",
      "VVGPYGLLQPFSDAMKLFIKEPLRPNFSSTALFMIAPTLA ENSJJAP00000000001.1\n",
      "IVGPYGILQPFADAMKLFMKEPMRPLTTSMSLFIIAPTLS ENSMUSP00000080991.2\n",
      "IVGPYGILQPFADAMKLFMKEPMRPLTTSMSLFIIAPTLS ENSRNOP00000049172.4\n",
      "IVGPYGILQPFADAMKLFIKEPLRPLATSSTLFIIAPTLS ENSMOCP00000000001.1\n",
      "IVGPYGILQPFADAMKLFIKEPLRPLATSTTLFILAPTLS ENSCGRP00000000001.1\n",
      "IVGPYGILQPFADAMKLFIKEPLRPLATSTSLFIIAPTLS ENSMAUP00000000001.1\n"
     ]
    }
   ],
   "source": [
    "from Bio import AlignIO\n",
    "jj_aln = AlignIO.read(open('aln/%s.aln' % 'ENSJJAG00000000006'), format='clustal')\n",
    "seq_id_list = [i.id.split('.')[0] for i in jj_aln._records]\n",
    "jj_aln_idx = seq_id_list.index('ENSJJAP00000000001')\n",
    "jj_length = len(jj_aln[jj_aln_idx])\n",
    "jj_pos = 58\n",
    "region_start = max(0, jj_pos-20)\n",
    "region_end = min(jj_length, jj_pos + 20)\n",
    "print(jj_aln[:, region_start:region_end])\n",
    "# change jj_aln to array, use as symble\n",
    "# def aln_color(aln): return color\n",
    "# plot anno heatmap\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(6, 40)"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import numpy as np\n",
    "from collections import Counter\n",
    "aln_array = np.array(jj_aln[:, region_start:region_end])\n",
    "aln_array.shape\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [],
   "source": [
    "color_scheme_Z = {\n",
    "    # Hydrophobic\n",
    "    'W': 0.0,\n",
    "    'L': 0.0,\n",
    "    'V': 0.0,\n",
    "    'I': 0.0,\n",
    "    'M': 0.0,\n",
    "    'F': 0.0,\n",
    "    'A': 0.0,\n",
    "    'C': 0.0,\n",
    "    # Positive charge\n",
    "    'K': 0.1,\n",
    "    'R': 0.1,\n",
    "    # Negative charge\n",
    "    'E': 0.2,\n",
    "    'D': 0.2,\n",
    "    # Polar\n",
    "    'N': 0.3,\n",
    "    'Q': 0.3,\n",
    "    'S': 0.3,\n",
    "    'T': 0.3,\n",
    "    # Cysteines\n",
    "    'C': 0.4,\n",
    "    # Glycines\n",
    "    'G': 0.5,\n",
    "    # Prolines\n",
    "    'P': 0.6,\n",
    "    # Aromatic\n",
    "    'H': 0.7,\n",
    "    'Y': 0.7,\n",
    "    # Other\n",
    "    '-': 1.0,\n",
    "}\n",
    "\n",
    "colorscale=[[0.0, 'rgb(0, 64, 255)'], [0.1, 'rgb(255, 0, 0)'], \n",
    "            [0.2, 'rgb(191, 0, 255)'], [0.3, 'rgb(0, 255, 0)'], \n",
    "            [0.4, 'rgb(255, 0, 255)'], [0.5, 'rgb(255, 128, 0)'],\n",
    "            [0.6, 'rgb(255, 255, 0)'], [0.7, 'rgb(0, 191, 255)'],\n",
    "            [1.0, 'rgb(0, 0, 0)']]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[[1.0,\n",
       "  0.0,\n",
       "  0.5,\n",
       "  0.6,\n",
       "  0.7,\n",
       "  0.5,\n",
       "  1.0,\n",
       "  0.0,\n",
       "  0.3,\n",
       "  0.6,\n",
       "  0.0,\n",
       "  1.0,\n",
       "  0.2,\n",
       "  0.0,\n",
       "  0.0,\n",
       "  0.1,\n",
       "  0.0,\n",
       "  0.0,\n",
       "  0.0,\n",
       "  0.1,\n",
       "  0.2,\n",
       "  0.6,\n",
       "  0.0,\n",
       "  0.1,\n",
       "  0.6,\n",
       "  1.0,\n",
       "  1.0,\n",
       "  1.0,\n",
       "  0.3,\n",
       "  1.0,\n",
       "  1.0,\n",
       "  0.0,\n",
       "  0.0,\n",
       "  1.0,\n",
       "  0.0,\n",
       "  0.0,\n",
       "  0.6,\n",
       "  0.3,\n",
       "  0.0,\n",
       "  1.0],\n",
       " [0.0,\n",
       "  0.0,\n",
       "  0.5,\n",
       "  0.6,\n",
       "  0.7,\n",
       "  0.5,\n",
       "  0.0,\n",
       "  0.0,\n",
       "  0.3,\n",
       "  0.6,\n",
       "  0.0,\n",
       "  0.0,\n",
       "  0.2,\n",
       "  0.0,\n",
       "  0.0,\n",
       "  0.1,\n",
       "  0.0,\n",
       "  0.0,\n",
       "  1.0,\n",
       "  0.1,\n",
       "  0.2,\n",
       "  0.6,\n",
       "  1.0,\n",
       "  0.1,\n",
       "  0.6,\n",
       "  0.0,\n",
       "  1.0,\n",
       "  0.3,\n",
       "  0.3,\n",
       "  1.0,\n",
       "  1.0,\n",
       "  0.0,\n",
       "  0.0,\n",
       "  0.0,\n",
       "  0.0,\n",
       "  0.0,\n",
       "  0.6,\n",
       "  0.3,\n",
       "  0.0,\n",
       "  0.3],\n",
       " [0.0,\n",
       "  0.0,\n",
       "  0.5,\n",
       "  0.6,\n",
       "  0.7,\n",
       "  0.5,\n",
       "  0.0,\n",
       "  0.0,\n",
       "  0.3,\n",
       "  0.6,\n",
       "  0.0,\n",
       "  0.0,\n",
       "  0.2,\n",
       "  0.0,\n",
       "  0.0,\n",
       "  0.1,\n",
       "  0.0,\n",
       "  0.0,\n",
       "  1.0,\n",
       "  0.1,\n",
       "  0.2,\n",
       "  0.6,\n",
       "  1.0,\n",
       "  0.1,\n",
       "  0.6,\n",
       "  0.0,\n",
       "  1.0,\n",
       "  0.3,\n",
       "  0.3,\n",
       "  1.0,\n",
       "  1.0,\n",
       "  0.0,\n",
       "  0.0,\n",
       "  0.0,\n",
       "  0.0,\n",
       "  0.0,\n",
       "  0.6,\n",
       "  0.3,\n",
       "  0.0,\n",
       "  0.3],\n",
       " [0.0,\n",
       "  0.0,\n",
       "  0.5,\n",
       "  0.6,\n",
       "  0.7,\n",
       "  0.5,\n",
       "  0.0,\n",
       "  0.0,\n",
       "  0.3,\n",
       "  0.6,\n",
       "  0.0,\n",
       "  0.0,\n",
       "  0.2,\n",
       "  0.0,\n",
       "  0.0,\n",
       "  0.1,\n",
       "  0.0,\n",
       "  0.0,\n",
       "  0.0,\n",
       "  0.1,\n",
       "  0.2,\n",
       "  0.6,\n",
       "  0.0,\n",
       "  0.1,\n",
       "  0.6,\n",
       "  0.0,\n",
       "  1.0,\n",
       "  0.3,\n",
       "  0.3,\n",
       "  1.0,\n",
       "  1.0,\n",
       "  0.0,\n",
       "  0.0,\n",
       "  0.0,\n",
       "  0.0,\n",
       "  0.0,\n",
       "  0.6,\n",
       "  0.3,\n",
       "  0.0,\n",
       "  0.3],\n",
       " [0.0,\n",
       "  0.0,\n",
       "  0.5,\n",
       "  0.6,\n",
       "  0.7,\n",
       "  0.5,\n",
       "  0.0,\n",
       "  0.0,\n",
       "  0.3,\n",
       "  0.6,\n",
       "  0.0,\n",
       "  0.0,\n",
       "  0.2,\n",
       "  0.0,\n",
       "  0.0,\n",
       "  0.1,\n",
       "  0.0,\n",
       "  0.0,\n",
       "  0.0,\n",
       "  0.1,\n",
       "  0.2,\n",
       "  0.6,\n",
       "  0.0,\n",
       "  0.1,\n",
       "  0.6,\n",
       "  0.0,\n",
       "  1.0,\n",
       "  0.3,\n",
       "  0.3,\n",
       "  1.0,\n",
       "  1.0,\n",
       "  0.0,\n",
       "  0.0,\n",
       "  0.0,\n",
       "  1.0,\n",
       "  0.0,\n",
       "  0.6,\n",
       "  0.3,\n",
       "  0.0,\n",
       "  0.3],\n",
       " [0.0,\n",
       "  0.0,\n",
       "  0.5,\n",
       "  0.6,\n",
       "  0.7,\n",
       "  0.5,\n",
       "  0.0,\n",
       "  0.0,\n",
       "  0.3,\n",
       "  0.6,\n",
       "  0.0,\n",
       "  0.0,\n",
       "  0.2,\n",
       "  0.0,\n",
       "  0.0,\n",
       "  0.1,\n",
       "  0.0,\n",
       "  0.0,\n",
       "  0.0,\n",
       "  0.1,\n",
       "  0.2,\n",
       "  0.6,\n",
       "  0.0,\n",
       "  0.1,\n",
       "  0.6,\n",
       "  0.0,\n",
       "  1.0,\n",
       "  0.3,\n",
       "  0.3,\n",
       "  1.0,\n",
       "  1.0,\n",
       "  0.0,\n",
       "  0.0,\n",
       "  0.0,\n",
       "  0.0,\n",
       "  0.0,\n",
       "  0.6,\n",
       "  0.3,\n",
       "  0.0,\n",
       "  0.3]]"
      ]
     },
     "execution_count": 38,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def aln_color_z(aln, color_code):\n",
    "    shape = aln.shape\n",
    "    coloring_count_threshold = int(min(shape[0]-1, shape[0]*0.8))\n",
    "    color_cols = []\n",
    "    for i in range(shape[1]):\n",
    "        cur_col = aln[:, i]\n",
    "        cur_col_count = Counter(cur_col)\n",
    "        if len(sorted(cur_col_count.elements())) == 1: # all the same\n",
    "            coloring = [cur_col_count.elements()]\n",
    "        else:\n",
    "            coloring = []\n",
    "            for j in cur_col_count.most_common(3):\n",
    "                if j[1] >= coloring_count_threshold:\n",
    "                    coloring.append(j[0])\n",
    "        cur_color = []\n",
    "        for aa in cur_col:\n",
    "            if aa in coloring and aa != '-':\n",
    "                cur_color.append(color_code[aa])\n",
    "            else:\n",
    "                cur_color.append(color_code['-'])\n",
    "        color_cols.append(cur_color)\n",
    "    z = np.array(color_cols).T.tolist()\n",
    "    return z\n",
    "aln_color_z(aln_array, color_code=color_scheme_Z)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.5.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
